File: web-cat-src/Eclipse/CxxTest/cxxtest-core/src/net/sf/webcat/eclipse/cxxtest/internal/generator/TestRunnerGenerator.java
/*==========================================================================*\
 |  $Id: TestRunnerGenerator.java,v 1.5 2009/09/13 16:13:15 aallowat Exp $
 |*-------------------------------------------------------------------------*|
 |  Copyright (C) 2006-2009 Virginia Tech 
 |
 |	This file is part of Web-CAT Eclipse Plugins.
 |
 |	Web-CAT is free software; you can redistribute it and/or modify
 |	it under the terms of the GNU General Public License as published by
 |	the Free Software Foundation; either version 2 of the License, or
 |	(at your option) any later version.
 |
 |	Web-CAT is distributed in the hope that it will be useful,
 |	but WITHOUT ANY WARRANTY; without even the implied warranty of
 |	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 |	GNU General Public License for more details.
 |
 |	You should have received a copy of the GNU General Public License
 |	along with Web-CAT; if not, see <http://www.gnu.org/licenses/>.
\*==========================================================================*/

package net.sf.webcat.eclipse.cxxtest.internal.generator;

import java.io.FileWriter;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.Writer;
import java.util.ArrayList;
import java.util.Hashtable;
import java.util.Map;

import net.sf.webcat.eclipse.cxxtest.CxxTestPlugin;
import net.sf.webcat.eclipse.cxxtest.ICxxTestConstants;

import org.antlr.stringtemplate.AutoIndentWriter;
import org.antlr.stringtemplate.StringTemplate;
import org.antlr.stringtemplate.StringTemplateGroup;
import org.antlr.stringtemplate.language.AngleBracketTemplateLexer;
import org.eclipse.cdt.core.model.ICProject;
import org.eclipse.core.runtime.FileLocator;
import org.eclipse.core.runtime.Path;

/**
 * 
 * @author  Tony Allevato (Virginia Tech Computer Science)
 * @author  latest changes by: $Author: aallowat $
 * @version $Revision: 1.5 $ $Date: 2009/09/13 16:13:15 $
 */
public class TestRunnerGenerator
{
    /**
     * Instantiates an instance of the CxxTestDriverGenerator for the specified
     * project and test suite collection.
     * 
     * @param project
     *            the ICProject associated with this generator
     * @param path
     *            the path of the source file to be generated
     * @param suites
     *            the collection of test suites to be generated
     * 
     * @throws IOException
     *             if an I/O error occurs during generation
     */
    public TestRunnerGenerator(ICProject project, String path,
    		TestSuiteCollection suites, Map<String, Boolean> testsToRun)
    throws IOException
    {
        this.suites = suites;

        // Create a proxy object to manage the tests to run. Any tests
        // not in this map are assumed to be true (so that if tests
        // have been added, but not refreshed in the tool window, they
        // will be run until they are explicitly disabled).

        this.testsToRunProxy = new TestsToRunProxy();

        // Load the template from the embedded assembly resources.

        InputStream stream = FileLocator.openStream(
        		CxxTestPlugin.getDefault().getBundle(),
        		new Path(RunnerTemplateResourcePath), true);

        StringTemplateGroup templateGroup = new StringTemplateGroup(
                new InputStreamReader(stream), AngleBracketTemplateLexer.class);

        templateGroup.registerRenderer(String.class,
                new TestRunnerStringRenderer(path));

        template = templateGroup.getInstanceOf("runAllTestsFile"); //$NON-NLS-1$

        // Initialize the options that will be passed into the template.

        options = new Hashtable<String, Object>();
        options.put("platformIsMSVC", false); //$NON-NLS-1$
        options.put("trapSignals", true); //$NON-NLS-1$
        options.put("traceStack", true); //$NON-NLS-1$
        options.put("noStaticInit", true); //$NON-NLS-1$
        options.put("root", true); //$NON-NLS-1$
        options.put("part", false); //$NON-NLS-1$
        options.put("abortOnFail", true); //$NON-NLS-1$
        options.put("mainProvided", suites.doesMainFunctionExist()); //$NON-NLS-1$
        options.put("testResultsFilename", ICxxTestConstants.TEST_RESULTS_FILE); //$NON-NLS-1$
        options.put("testsToRun", testsToRunProxy); //$NON-NLS-1$

        ArrayList<String> listeners = new ArrayList<String>();
        listeners.add("XmlStdioPrinter"); //$NON-NLS-1$
        options.put("listeners", listeners); //$NON-NLS-1$
        
        try
        {
            writer = new FileWriter(path);
        }
        catch (IOException e)
        {
            e.printStackTrace();
        }
    }


	public boolean isTrackingHeap()
	{
		return trackHeap;
	}

	public void setTrackHeap(boolean value)
	{
		trackHeap = value;
	}

	public boolean isTrappingSignals()
	{
		return trapSignals;
	}

	public void setTrapSignals(boolean value)
	{
		options.put("trapSignals", value);
		trapSignals = value;
	}

	public boolean isTracingStack()
	{
		return traceStack;
	}
	
	public void setTraceStack(boolean value)
	{
		options.put("traceStack", value);
		traceStack = value;
	}

	public String[] getExtraIncludes()
	{
		return extraIncludes;
	}

	public void setExtraIncludes(String[] files)
	{
		extraIncludes = files;
	}
	
	public String[] getPossibleTestFiles()
	{
		return possibleTestFiles;
	}

	public void setPossibleTestFiles(String[] files)
	{
		possibleTestFiles = files;
	}
	
	public void generate()
    {
        template.setAttribute("options", options); //$NON-NLS-1$
        template.setAttribute("suites", suites.getSuites()); //$NON-NLS-1$

        if (possibleTestFiles != null && possibleTestFiles.length > 0)
        {
        	template.setAttribute("possibleTestFiles", possibleTestFiles); //$NON-NLS-1$
        }

        if (extraIncludes != null && extraIncludes.length > 0)
        {
        	options.put("extraIncludes", extraIncludes); //$NON-NLS-1$
        }

        try
        {
            template.write(new AutoIndentWriter(writer));
            writer.close();
        }
        catch (IOException e)
        {
            e.printStackTrace();
        }
    }


    private class TestsToRunProxy extends Hashtable<String, Boolean>
    {
        private static final long serialVersionUID = 1L;

        public boolean containsKey(Object key)
        {
            return true;
        }
        
        public Boolean get(Object key)
        {
            return true;
        }
    }


    private static final String RunnerTemplateResourcePath =
    	"/generator-templates/runner.stg"; //$NON-NLS-1$

    private TestSuiteCollection suites;
    private TestsToRunProxy testsToRunProxy;
    private Hashtable<String, Object> options;
    private StringTemplate template;
    private Writer writer;
    
	private boolean trackHeap;
	private boolean trapSignals;
	private boolean traceStack;

	private String[] possibleTestFiles;
	private String[] extraIncludes;
}

File: web-cat-src/VisualStudio/CxxTestForVS/CxxTestPackage/Templating/TestRunnerGenerator.cs
/*==========================================================================*\
 |  $Id: TestRunnerGenerator.cs,v 1.1 2008/06/02 23:27:40 aallowat Exp $
 |*-------------------------------------------------------------------------*|
 |  Copyright (C) 2008 Virginia Tech
 |
 |  This file is part of the Web-CAT CxxTest integration package for Visual
 |	Studio.NET.
 |
 |  Web-CAT is free software; you can redistribute it and/or modify
 |  it under the terms of the GNU Affero General Public License as published
 |  by the Free Software Foundation; either version 3 of the License, or
 |  (at your option) any later version.
 |
 |  Web-CAT is distributed in the hope that it will be useful,
 |  but WITHOUT ANY WARRANTY; without even the implied warranty of
 |  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 |  GNU General Public License for more details.
 |
 |  You should have received a copy of the GNU Affero General Public License
 |  along with Web-CAT; if not, see <http://www.gnu.org/licenses/>.
\*==========================================================================*/

using System;
using System.Collections.Generic;
using System.Text;
using WebCAT.CxxTest.VisualStudio.Model;
using System.IO;
using EnvDTE;
using System.Reflection;
using Microsoft.VisualStudio.VCCodeModel;
using WebCAT.CxxTest.VisualStudio.Utility;
using Antlr.StringTemplate;
using Antlr.StringTemplate.Language;
using System.Collections;
using System.Collections.Specialized;

namespace WebCAT.CxxTest.VisualStudio.Templating
{
	class TestRunnerGenerator
	{
		/**
		 * Instantiates an instance of the CxxTestDriverGenerator for the
		 * specified project and test suite collection.
		 * 
		 * @param project the ICProject associated with this generator
		 * @param path the path of the source file to be generated
		 * @param suites the collection of test suites to be generated
		 * 
		 * @throws IOException if an I/O error occurs during generation
		 */
		public TestRunnerGenerator(string path, TestSuiteCollection suites,
			Dictionary<string, bool> testsToRun)
		{
			this.suites = suites;
			this.runnerPath = path;

			// Create a proxy object to manage the tests to run. Any tests
			// not in this map are assumed to be true (so that if tests
			// have been added, but not refreshed in the tool window, they
			// will be run until they are explicitly disabled).

			this.testsToRunProxy = new TestsToRunProxy(testsToRun);

			// Load the template from the embedded assembly resources.

			Stream stream =
				Assembly.GetExecutingAssembly().GetManifestResourceStream(
				RunnerTemplateResourcePath);

			StringTemplateGroup templateGroup = new StringTemplateGroup(
				new StreamReader(stream), typeof(AngleBracketTemplateLexer));

			templateGroup.RegisterAttributeRenderer(typeof(string),
				new TestRunnerStringRenderer(path));

			template = templateGroup.GetInstanceOf("runAllTestsFile");

			// Initialize the options that will be passed into the template.

			options = new Hashtable();
			options["platformIsMSVC"] = true;
			options["trapSignals"] = true;
			options["traceStack"] = true;
			options["noStaticInit"] = true;
			options["root"] = true;
			options["part"] = false;
			options["abortOnFail"] = true;
			options["longLongType"] = null;
			options["mainProvided"] = suites.MainFunctionExists;
			options["runner"] = "XmlStdioPrinter";
			options["xmlOutput"] = true;
			options["testResultsFilename"] = Constants.TestResultsFilename;
			options["testsToRun"] = testsToRunProxy;

			writer = File.CreateText(path);
		}
		
		public void Generate()
		{
			template.SetAttribute("options", options);
			template.SetAttribute("suites",
				new List<TestSuite>(suites.Suites));

			if (suites.PossibleTestFiles.Length > 0)
			{
				template.SetAttribute("possibleTestFiles",
					suites.PossibleTestFiles);
			}

			template.Write(new AutoIndentWriter(writer));

			writer.Close();
		}


		private class TestsToRunProxy
		{
			public TestsToRunProxy(IDictionary<string, bool> testsToRun)
			{
				this.testsToRun = testsToRun;
			}

			public bool this[string test]
			{
				get
				{
					if (testsToRun == null)
						return false;
					else if (!testsToRun.ContainsKey(test))
						return true;
					else
						return testsToRun[test];
				}
			}

			private IDictionary<string, bool> testsToRun;
		}

		private const string RunnerTemplateResourcePath =
			"WebCAT.CxxTest.VisualStudio.Resources.Templates.runner.stg";

		private string runnerPath;
		private TestSuiteCollection suites;
		private TestsToRunProxy testsToRunProxy;
		private Hashtable options;
		private StringTemplate template;
		private TextWriter writer;
	}
}
